bindings/hnsw_bindings.cpp

```diff
@@ -799,6 +827,79 @@ template <typename dist_t, typename data_t = float> class Index {
   size_t getMaxElements() const { return appr_alg->max_elements_; }
 
   size_t getCurrentCount() const { return appr_alg->cur_element_count; }
+
+  void loadPQ(const py::object pq_abstract) {
+    int exist_attr = 1, attr_correct = 1;
+    PyObject *pq_raw_ptr = pq_abstract.ptr();
+    exist_attr *= PyObject_HasAttrString(pq_raw_ptr, "encode");
+    exist_attr *= PyObject_HasAttrString(pq_raw_ptr, "get_codebook");
+    exist_attr *= PyObject_HasAttrString(pq_raw_ptr, "get_subspace_splitting");
+    if (exist_attr <= 0) {
+      throw py::index_error(
+          "PQ class should at least have the following attributes:\n"
+          "(encode, get_codebook, get_subspace_splitting)");
+    }
+    attr_correct *=
+        PyMethod_Check(PyObject_GetAttrString(pq_raw_ptr, "encode"));
+    attr_correct *=
+        PyMethod_Check(PyObject_GetAttrString(pq_raw_ptr, "get_codebook"));
+    attr_correct *= PyMethod_Check(
+        PyObject_GetAttrString(pq_raw_ptr, "get_subspace_splitting"));
+    if (attr_correct <= 0) {
+      throw py::attribute_error(
+          "PQ class have at least one of the following attributes' type "
+          "INCORRECT:\n(encode: <bounded method>,\n codebook: "
+          "<bounded method>,\n get_subspace_splitting: <bounded method>)");
+    }
+
+    this->pq_enable = true;
+    this->pq_codec = pq_abstract;
+
+    py::tuple subspaces_param = pq_abstract.attr("get_subspace_splitting")();
+    this->pq_n_subvectors = py::cast<size_t>(subspaces_param[0]);
+    this->pq_n_clusters = py::cast<size_t>(subspaces_param[1]);
+    this->pq_d_subvector = py::cast<size_t>(subspaces_param[2]);
+
+    size_t pq_total_dims = (this->pq_n_subvectors * this->pq_d_subvector);
+    if (this->dim != pq_total_dims) {
+      throw py::value_error(
+          "Initialization Error, expect HNSW.dim == "
+          "PQ.n_subvector*PQ.d_subvector, but got:\n"
+          "HNSW.dim =" +
+          std::to_string(this->dim) +
+          ", PQ.n_subvector*PQ.d_subvector=" + std::to_string(pq_total_dims));
+    }
+    // reading codebook into float buffer
+    py::array_t<dist_t, py::array::c_style | py::array::forcecast> items(
+        pq_abstract.attr("get_codebook")());
+    auto buffer = items.request();
+    if (buffer.ndim != 3 || buffer.shape[0] != pq_n_subvectors ||
+        buffer.shape[1] != pq_n_clusters || buffer.shape[2] != pq_d_subvector) {
+      py::print("Expect the codebook with shape (", pq_n_subvectors,
+                pq_n_clusters, pq_d_subvector, "seq"_a = ",");
+      py::print(" but got shape ", buffer.shape);
+      throw py::attribute_error(
+          "PQ class returning the codebook with wrong dimension");
+    }
+    std::shared_ptr<float *> codebook_buffer =
```

Comment:
```suggestion
    std::shared_ptr<float> codebook_buffer =
```

Reply:
the template type is the type, u do not want a shared pointer to a float pointer, but a shared poiinter to a float.

Reply:
But now that I see it, I am not sure this is what u want, what u would like is to have an `std::array<float>` instead of a `float*`. Like this u can use this `std::array` as an object and forget about the pointers, and just pass it around as const reference

Reply:
If the `codebook` has a known size u should use `std::array` to handle the ownership of the underlying buffer in a safe way

Reply:
I made a mistake about the `shared_ptr`'s type here, it should be `float`.
May I ask when you suggested `std::array`, what's the solution in your mind? Do you mean we should copy the codebook from the buffer?